README.md
# TP-Link / Kasa Smart Home Device SmartThings Integration
There is only one version supporting three integrations:

a.  Kasa Account.  Integration via the user's Kasa Account through the Cloud.  This uses a Smart Application for installation, device communications, and device management.

b.  Node Applet.  Smart Application integration via a home wifi Node.JS bridge (pc, fire tablet, android device, Raspberry Pi).  The Smart Application is used (with user entry of bridge IP) to install and manage the devices.  Especially useful in the new SmartThings phone app since it allows entry of user preferences via that app.

c.  Manual Node Installation.  Traditional Hub installation.  Does not use a Smart Application.

# All three versions are compatible with the new SmartThings phone application.

# Installation Prequisites:

A SmartThings Hub, IDE Account and SmartThings Classic are required for all original installations.  After installation, you may (if desired) transition to the new SmartThings phone app.

a.  Kasa Account.  (1) Kasa Account, (2) TP-Link devices in remote mode.

b.  Node Applet.  (1) Node.js Bridge, (2) Static IP address for Bridge (recommended for all devices).

c.  Manual Node Installation.  (1)  Node.js Bridge, (2) Static IP addresses for the bridge and all devices.

# Installation Instructions

(For interaction with the IDE, look in the SmartThings Community Forum.)

1.  Copy device handlers and in

# Installation (Must use the SmartThings Classic App)
Installation instruction can be found in the Documentation Folders.  These are step-by-step and are for users new to SmartThings.  However, they have not been modified from the previous version yet.

a.  Install the relevant installation file(s):
    
    1.  Go to the MyDeviceHandler page on the IDE, select 'Create New Device Handler'  Select 'From Code'.
    
    2.  Copy the contents of the relevant file in the 'Device Handler' folder herein.  Paste into the IDE page.

    3.  Select "Create".  Select 'Publish' 'For Me'.
    
    4.  For 'Kasa Account' or 'Node Applet' installations, go to the My SmartAps page on the IDE.  Select 'New SmartApp'.  Select 'From Code'.
    
    5.  Copy the contents of 'TP-Link SmartThings Manager.groovy' in the 'Service Manager' folder herein.  Paste into the IDE page.
    
    6.  Seleect 'Create'.  Select 'Publish' 'For Me'

b.  Installation of 'Kasa Account' or 'Node Applet' integrations.

    1.  From the SmartThings Classic app on your phone, select 'SmartApps', then 'Add a SmartApp'.  Select 'My Apps' at bottom of next page.  Select 'TP-Link SmartThings Manager' from the 'My Apps' page.
    
    2.  'Select Installation Type'.  Tap for the selection of 'Kasa Account' or 'Node Applet'.  Once you select, the program will land on one of two pages
    
    3.  'Kasa Account'.  Enter your Username and Password.  Once both are entered (right or wrong), you will be directet to select in "Install a Device to Continue.  That will take you to the 'Device Installation Page.  Follow prompts to install the devices.
    
    4.  'Node Applet'. Assure the Node.js Applet is running.  Enter the device IP (example:  192.168.1.199) for your bridge.  You will see an error until the system has time to actually detect devices.  Then follow prompts to add devices.

c.  Installation of Manual Node Installation

  1.  Start the Node.js Applet.  Go to the IDE, "My Devices'.  Select 'New Device'.
  
  2.  Fill-out the form fields 'Name', 'Label', 'Device Network ID'.
  
  3.  From the pull-down 'Type', select the appropriate device type (handler).  Select 'Location' (your hub).
  
  4.  In 'My devices or on the phone (Classic App) for each device, select preferences.  Inter InstallType, Bridge IP, and Device IP and save.
  
  5.  You can now use either phone app to control the device.  Prefernces can only be updated on the IDE or the classic app.

# Version 3.5 Update - 10/31/2018

The Files have been updated to Version 3.5.

Note:  In the Node Applet and Kasa Account installation, DO NOT CHANGE preferences BridgeIP, DeviceIP, nor InstallType.
Device Handler Changes

1.  Single device handler for Hub or Cloud (now 'Node Applet' or 'Kasa Account').  Added preference installType to accommodate.

2.  Update to work with the new SmartThings Application.  Note that this application has some limits:

    a.  No preferences page for devices.
  
    b.  Functionality limited to Samsung selected (no energy monitor display, no Circadian Mode control).
 
Smart Application Update:

1.  Smart App now supports both 'Kasa Account' and 'Node Applet'.

2.  Added Remove Selected Devices and Update Selected Device Preference capabilities.

3.  Better Maintenance functions:  Manage Kasa Acct Settings, Change Bridge IP Address.

4.  Can actually track Device IPs, therefore (although recommended) static deviceIPs are not required.  Static Bridge IP remains required.

# Update.  Unless you need to, not necessary at this time.  Will cause problems later when managing or installing new devices.
Update is possible.  It has not been extensively tested due to the complexity.  If the update fails, try a clean re-installation.

1.  The new device handlers will work with the existing Smart App (Service Manager) except if you try to ADD DEVICES.

2.  The new Smart App requires the new Device Handlers if you are going to perform any functions in the smartapp (it relies on the new naming conventions.


data/changelog.txt
##### TP-Link SmartThings Manager #####
2018-10-15 - Update to Version 3.6.0
a. Added 3 new page/icons (@ramiran2)
b. Improved UI Elements (@ramiran2)
f. Added Recommended Options (@ramiran2)

2018-10-13 - Update to Version 3.5.0
a. Added a new page/icons (@ramiran2)
b. Added icons to paragraph (@ramiran2)
c. Added Lighting Transition Time control (@ramiran2)
d. Added Device Refresh Rate control (@ramiran2)
e. Improved UI Elements (@ramiran2)
f. Added Recommended Options (@ramiran2)

2018-10-10 - Update to Version 3.4.0
a. Improved UI Elements (@ramiran2)
b. Improved logic (@ramiran2)
c. Added token page (@ramiran2)
d. Added Status of token to token page (@ramiran2)

2018-10-09 - Update to Version 3.2.5
a. Improved developer testing page (@ramiran2)
b. Improved error messages (@ramiran2)

2018-10-08 - Update to Version 3.2.0
a. Improved developer testing mode (@ramiran2)
b. Added developer testing page (@ramiran2)

2018-10-08 - Update to Version 3.1.3
a. Small tweeks to code (@ramiran2)

2018-10-08 - Update to Version 3.1.2
a. Removed developer page from login page as a selectable option as there is no need for it (@ramiran2)

2018-10-08 - Update to Version 3.1.1
a. Fixed a bug that sent users to the main page instead of the login page if username or password is null (@ramiran2)

2018-10-04 - Update to Version 3.1.0
a. Improved UI Elements with other small changes (@ramiran2)
b. Added a changelog page (@ramiran2)
c. Added a uninstall page (@ramiran2)
d. Added a about page (@ramiran2)
e. Improved logic (@ramiran2)

2018-10-01 - Update to Version 2.9.0
a. Improved UI Elements with other small changes (@ramiran2)
b. Added a developer page (@ramiran2)

2018-09-28 - Update to Version 2.8.0
a. Improved UI Elements with other small changes (@ramiran2)
b. Added a login page (@ramiran2)
c. Updated Driver Version Variables (@ramiran2)
d. Added a New Device Handler (@ramiran2)

2018-09-27 - Update to Version 2.7.0
a. Improved UI Elements with other small changes (@ramiran2)
b. Updated for new Device Handlers (@ramiran2)
c. Add Support for the new Smart Thing Application (@ramiran2)

2018-08-28 - Update to Version 2.6.0
a. Improved UI Elements with other small changes (@ramiran2)

2018-08-27 - Update to Version 2.5.0
a. Improved UI Elements with other large changes (@ramiran2)

2018-08-22 - Update to Version 2.2.0
a. Improved UI Elements and updated the app logo plus other small changes (@ramiran2)

2018-08-11 - Update to Version 2.1.0
a. Updated for Added Support for update from a repo on smartthings website (@ramiran2)
b. Improved app name (@ramiran2)
c. Added app version (@ramiran2)

2018-01-31 - Update to Version 2.0.0
a. Updated for new release of Device Handlers (@DaveGut)

##### TP-Link Smart Plug #####
2018-10-16 - Update to Version 3.2.0
a. Added modifcation from original repo (@DaveGut)
b. Improved Lighting Transition Time / Device Refresh Rate control (@DaveGut)
c. Removed code from switches (@ramiran2)
d. Added Energy Monitor Plug to device selector (@ramiran2)

2018-10-08 - Update to Version 3.1.3
a. Small tweeks to code (@ramiran2)
b. Changed icons in clasic mode (@ramiran2)

2018-10-04 - Update to Version 3.1.0
a. Added Device Health Check via Ping (@ramiran2)
b. Small tweeks to code (@ramiran2)

2018-09-28 - Update to Version 2.9.0
a. Added Device Health Check (@ramiran2)
b. Tweek to support for new tp-link devices (@DaveGut)
c. Added support for the new Smart Things app (@ramiran2)
d. Seperated Device Handler of Smart Plug/Switch (@ramiran2)

2018-08-11 - Update to Version 2.5.0
a. Added Support for update from a repo on smartthings website (@ramiran2)
b. Improved driver names (@ramiran2)
c. Added driver version (@ramiran2)

2018-01-31 - Update to Version 2.0.0
a. Common file content for all bulb implementations, using separate files by model only. (@DaveGut)
b. User file-internal selection of Energy Monitor function enabling. (@DaveGut)

##### TP-Link Smart Energy Monitor Plug #####
2018-10-16 - Update to Version 3.2.0
a. Added modifcation from original repo (@DaveGut)
b. Improved Lighting Transition Time / Device Refresh Rate control (@DaveGut)

2018-10-08 - Update to Version 3.1.3
a. Small tweeks to code (@ramiran2)
b. Changed icons in clasic mode (@ramiran2)

2018-10-04 - Update to Version 3.1.0
a. Added Device Health Check via Ping (@ramiran2)
b. Small tweeks to code (@ramiran2)

2018-09-28 - Update to Version 2.9.0
a. Added Device Health Check (@ramiran2)
b. Tweek to support for new tp-link devices (@DaveGut)
c. Added support for the new Smart Things app (@ramiran2)

2018-08-11 - Update to Version 2.5.0
a. Added Support for update from a repo on smartthings website (@ramiran2)
b. Improved driver names (@ramiran2)
c. Added driver version (@ramiran2)

2018-01-31 - Update to Version 2.0.0
a. Common file content for all bulb implementations, using separate files by model only. (@DaveGut)
b. User file-internal selection of Energy Monitor function enabling. (@DaveGut)

2018-02-17 - Update to Version 1.9.0
a. Updated Energy Monitor Functions (@DaveGut)
b. Allowed for full month collection in previous month (@DaveGut)
c. Cleaned-up algorithm to use Groovy date. (@DaveGut)

2018-02-19 - Update to Version 1.5.0
a. Completed Energy Monitor tuning (@DaveGut)
b. Fixed March 1, 2 issue where data would not be captured (@DaveGut)
c. Update remaining code. (@DaveGut)

2018-04-22 - Initial Release Version 1.0.0
a. Update setCurrentDate to eliminate error for some users. (@DaveGut)

##### TP-Link Smart Tunable White Bulb #####
2018-10-16 - Update to Version 3.2.0
a. Added modifcation from original repo (@DaveGut)
b. Improved Lighting Transition Time / Device Refresh Rate control (@DaveGut)

2018-10-08 - Update to Version 3.1.3
a. Small tweeks to code (@ramiran2)
b. Changed icons in clasic mode (@ramiran2)


NodeJS Files/Utilities/GetDeviceIPs.js
const dgram = require('dgram')
var os = require('os')


//	-------------------------------------------------------------------
console.log("List of TP-Link Devices with IP, MAC, Alias, and Model")
var interfaces = os.networkInterfaces()
for (var k in interfaces) {
	for (var k2 in interfaces[k]) {
		var address = interfaces[k][k2]
		if (address.family === 'IPv4' && !address.internal) {
			var bridgeIP = address.address
		}
	}
}
var bridgePort = 8090
var socket = dgram.createSocket('udp4')
var tplinkDeviceList = {}
setTimeout(displayDevices, 4000)
socket.on('error', function (err) {
	console.error('Client UDP error')
	console.trace(err)
	socket.close()
	emit('error', err)
}.bind())
socket.on('message', function (msg, rinfo) {
	var tplinkDevice = {}
	var device = JSON.parse(decrypt(msg).toString('ascii')).system.get_sysinfo
	if (device.mic_type == "IOT.SMARTBULB") {
		tplinkDNI = device.mic_mac
	} else {
		tplinkDNI = device.mac.replace(/:/g, "")
	}
	tplinkDevice['deviceIP'] = rinfo.address
	tplinkDevice['deviceModel'] = device.model
	tplinkDevice['deviceAlias'] = device.alias
	tplinkDevice['deviceId'] = device.deviceId
	tplinkDeviceList[tplinkDNI] = tplinkDevice
}.bind())
socket.bind(bridgePort, bridgeIP, () => {
	socket.setBroadcast(true)
})
var msgBuf = UdpEncrypt('{"system":{"get_sysinfo":{}}}')
socket.send(msgBuf, 0, msgBuf.length, 9999, '255.255.255.255')
//	-------------------------------------------------------------------
function displayDevices() {
	socket.close()
	console.log(tplinkDeviceList)
}
//	-------------------------------------------------------------------
function UdpEncrypt (input) {
	var buf = new Buffer(input.length)
	var key = 0xAB
	for (var i = 0; i < input.length; i++) {
		buf[i] = input.charCodeAt(i) ^ key
		key = buf[i]
	}
	return buf
}
//	-------------------------------------------------------------------
function TcpEncrypt (input) {
	var buf = Buffer.alloc(input.length)
	var key = 0xAB
	for (var i = 0; i < input.length; i++) {
		buf[i] = input.charCodeAt(i) ^ key
		key = buf[i]
	}
	var bufLength = Buffer.alloc(4)
	bufLength.writeUInt32BE(input.length, 0)
	return Buffer.concat([bufLength, buf], input.length + 4)
}
//	-------------------------------------------------------------------
function decrypt (input, firstKey) {
	var buf = Buffer.from(input); // node v6: Buffer.from(input)
	var key = 0x2B
	var nextKey
	for (var i = 0; i < buf.length; i++) {
		nextKey = buf[i]
		buf[i] = buf[i] ^ key
		key = nextKey
	}
	return buf
}

NodeJS Files/Utilities/HS110Simulator.js
/*
TP-LinkHub - Version 2.0

COMPATABILITY KEY:  HubVersion 2.0

This java script uses node.js functionality to provide a hub between SmartThings and TP-Link devices.  It works with the following TP-Link integrations:
a.	TP-Link Connect (including Discovery)
b.	TP-Link Smart Things Integration
c.	TP-Link Bridge (OPTIONAL)

01-31-2018	Release of Version 2 Hub
*/

//##### Options for this program ###################################
var logFile = "yes"	//	Set to no to disable error.log file.
var hubPort = 8082	//	Synched with Device Handlers.
//##################################################################

//---- Determine if old Node version, act accordingly -------------
console.log("Node.js Version Detected:   " + process.version)
var oldNode = "no"
if (process.version == "v6.0.0-pre") {
	oldNode ="yes"
	logFile = "no"
}

//---- Program set up and global variables -------------------------
var http = require('http')
var net = require('net')
var fs = require('fs')
var server = http.createServer(onRequest)

//---- Start the HTTP Server Listening to SmartThings --------------
server.listen(hubPort)
console.log("TP-Link Hub Console Log")
logResponse("\n\r" + new Date() + "\rTP-Link Hub Error Log")

//---- Command interface to Smart Things ---------------------------
function onRequest(request, response){
	var command = request.headers["command"]
	var deviceIP = request.headers["tplink-iot-ip"]
	var cmdRcvd = "\n\r" + new Date() + "\r\nIP: " + deviceIP + " sent command " + command
	console.log(" ")
	console.log(cmdRcvd)
	switch(command) {
		//---- (BridgeDH - Poll for Server APP ------------------
		case "pollServer":
			response.setHeader("cmd-response", "ok")
			response.end()
			var respMsg = "Server Poll response sent to SmartThings"
			console.log(respMsg)
		break

		//---- TP-Link Device Command ---------------------------
		case "deviceCommand":
			//----- Simulated power data Selection ----------------------
			var command = request.headers["tplink-command"]
			var action = request.headers["action"]
			if (action == "onWatts") {
				response.setHeader("action", "energyMeterResponse")
				response.setHeader("cmd-response", onWatts)
				response.end()
			} else if (action == "offWatts") {
				response.setHeader("action", "energyMeterResponse")
				response.setHeader("cmd-response", offWatts)
				response.end()
			} else {
			//----- Real data ----------------------
				processDeviceCommand(request, response)
			}
			break
	
		//---- Energy Meter Simulated Data Return ---------------
		case "emeterCmd":
			var action = request.headers["action"]
			var engrCmd = action.substring(0,3)
console.log("engrCmd = " + engrCmd)
			var engrData = action.substring(3)
console.log("engrData = " + engrData)
			if (engrCmd == "Con") {
				response.setHeader("action", "useTodayResponse")
			} else {
				response.setHeader("action", "engrStatsResponse")
			}
			var respData = ""
			switch(engrData) {
				case "DecWatts":
					respData = DecWatts
					break
				case "DecMWatts":
					respData = DecMWatts
					break
				case "JanWatts":
					respData = JanWatts
					break
				case "JanMWatts":
					respData = JanMWatts
					break
				case "FebWatts":
					respData = FebWatts
					break
				case "FebMWatts":
					respData = FebMWatts
					break
				case "MarWatts":
					respData = MarWatts
					break
				case "MarMWatts":
					respData = MarMWatts
					break
				case "Day1Watts":
					respData = Day1Watts
					break
				case "Day1MWatts":
					respData = Day1MWatts
					break
				case "Day2Watts":
					respData = Day2Watts
					break
				case "Day2MWatts":
					respData = Day2MWatts
					break
				case "Day3Watts":
					respData = Day3Watts
					break
				case "Day3MWatts":
					respData = Day3MWatts
					break
				default:
					break
			}
console.log(respData)
			response.setHeader("cmd-response", respData)
			response.end()
			break
		default:
			response.setHeader("cmd-response", "InvalidHubCmd")
			response.end()
			var respMsg = "#### Invalid Command ####"
			var respMsg = new Date() + "\n\r#### Invalid Command from IP" + deviceIP + " ####\n\r"
			console.log(respMsg)
			logResponse(respMsg)
	}
}

//---- Send deviceCommand and send response to SmartThings ---------
function processDeviceCommand(request, response) {
	var command = request.headers["tplink-command"]
	var deviceIP = request.headers["tplink-iot-ip"]
	var respMsg = "deviceCommand sending to IP: " + deviceIP + " Command: " + command
	console.log(respMsg)
	var action = request.headers["action"]
	response.setHeader("action", action)
	var socket = net.connect(9999, deviceIP)
	socket.setKeepAlive(false)
	socket.setTimeout(6000)
	socket.on('connect', () => {
		socket.write(TcpEncrypt(command))
	})
	socket.on('data', (data) => {
		socket.end()
		data = decrypt(data.slice(4)).toString('ascii')
		response.setHeader("cmd-response", data)
		response.end()
		var respMsg = "Command Response sent to SmartThings"
		console.log(respMsg)
	}).on('timeout', () => {
		response.setHeader("cmd-response", "TcpTimeout")
		response.end()
		socket.end()
		var respMsg = new Date() + "\n#### TCP Timeout in deviceCommand for IP: " + deviceIP + " ,command: " + command
		console.log(respMsg)
		logResponse(respMsg)
	}).on('error', (err) => {
		socket.end()
		var respMsg = new Date() + "\n#### Socket Error in deviceCommand for IP: " + deviceIP + " ,command: " + command
		console.log(respMsg)
		logResponse(respMsg)
	})
}

//----- Utility - Response Logging Function ------------------------
function logResponse(respMsg) {
	if (logFile == "yes") {
		fs.appendFileSync("error.log", "\r" + respMsg)
	}
}

//----- Utility - Encrypt TCP Commands to Devices ------------------
function TcpEncrypt(input) {
	if (oldNode == "no"){
		var buf = Buffer.alloc(input.length + 4)
	} else {
		var buf = new Buffer(input.length + 4)
	}
	buf[0] = null
	buf[1] = null
	buf[2] = null
	buf[3] = input.length
	var key = 0xAB
	for (var i = 4; i < input.length+4; i++) {
		buf[i] = input.charCodeAt(i-4) ^ key
		key = buf[i]
	}
	return buf
}

//----- Utility - Decrypt Returns from  Devices --------------------
function decrypt(input, firstKey) {
	if (oldNode == "no") {
		var buf = Buffer.from(input)
	} else {
		var buf = new Buffer(input)
	}
	var key = 0x2B
	var nextKey
	for (var i = 0; i < buf.length; i++) {
		nextKey = buf[i]
		buf[i] = buf[i] ^ key
		key = nextKey
	}
	return buf
}

//----- Simulator Data Value -----

//----- WATT / KW HR ====
var DecWatts = '{"emeter":{"get_daystat":{"day_list":[{"year":2017,"month":12,"day":19,"energy":3.77},{"year":2017,"month":12,"day":20,"energy":4.956},{"year":2017,"month":12,"day":21,"energy":0.481},{"year":2017,"month":12,"day":22,"energy":4.608},{"year":2017,"month":12,"day":23,"energy":1.538},{"year":2017,"month":12,"day":24,"energy":0.304},{"year":2017,"month":12,"day":25,"energy":0.976},{"year":2017,"month":12,"day":26,"energy":3.595},{"year":2017,"month":12,"day":27,"energy":2.125},{"year":2017,"month":12,"day":28,"energy":1.372},{"year":2017,"month":12,"day":29,"energy":1.403},{"year":2017,"month":12,"day":30,"energy":3.405},{"year":2017,"month":12,"day":31,"energy":2.924}],"err_code":0}}}'

var DecMWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2017,"month":12,"day":19,"energy_wh":3770},{"year":2017,"month":12,"day":20,"energy_wh":4956},{"year":2017,"month":12,"day":21,"energy_wh":481},{"year":2017,"month":12,"day":22,"energy_wh":4608},{"year":2017,"month":12,"day":23,"energy_wh":1538},{"year":2017,"month":12,"day":24,"energy_wh":304},{"year":2017,"month":12,"day":25,"energy_wh":976},{"year":2017,"month":12,"day":26,"energy_wh":3595},{"year":2017,"month":12,"day":27,"energy_wh":2125},{"year":2017,"month":12,"day":28,"energy_wh":1372},{"year":2017,"month":12,"day":29,"energy_wh":1403},{"year":2017,"month":12,"day":30,"energy_wh":3405},{"year":2017,"month":12,"day":31,"energy_wh":2924}],"err_code":0}}}'

var JanWatts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":1,"day":1,"energy":2.599},{"year":2018,"month":1,"day":2,"energy":3.392},{"year":2018,"month":1,"day":3,"energy":4.163},{"year":2018,"month":1,"day":4,"energy":1.207},{"year":2018,"month":1,"day":5,"energy":3.786},{"year":2018,"month":1,"day":6,"energy":2.17},{"year":2018,"month":1,"day":7,"energy":4.606},{"year":2018,"month":1,"day":8,"energy":1.56},{"year":2018,"month":1,"day":9,"energy":2.664},{"year":2018,"month":1,"day":10,"energy":4.567},{"year":2018,"month":1,"day":11,"energy":3.77},{"year":2018,"month":1,"day":12,"energy":4.956},{"year":2018,"month":1,"day":13,"energy":0.481},{"year":2018,"month":1,"day":14,"energy":4.608},{"year":2018,"month":1,"day":15,"energy":1.538},{"year":2018,"month":1,"day":16,"energy":0.304},{"year":2018,"month":1,"day":17,"energy":0.976},{"year":2018,"month":1,"day":18,"energy":3.595},{"year":2018,"month":1,"day":19,"energy":2.125},{"year":2018,"month":1,"day":20,"energy":1.372},{"year":2018,"month":1,"day":21,"energy":1.403},{"year":2018,"month":1,"day":22,"energy":3.405},{"year":2018,"month":1,"day":23,"energy":2.924},{"year":2018,"month":1,"day":24,"energy":2.599},{"year":2018,"month":1,"day":25,"energy":2.839},{"year":2018,"month":1,"day":26,"energy":1.329},{"year":2018,"month":1,"day":27,"energy":4.928},{"year":2018,"month":1,"day":28,"energy":4.282},{"year":2018,"month":1,"day":29,"energy":3.392},{"year":2018,"month":1,"day":30,"energy":4.163},{"year":2018,"month":1,"day":31,"energy":1.207}],"err_code":0}}}'

var JanMWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":1,"day":1,"energy_wh":2599},{"year":2018,"month":1,"day":2,"energy_wh":3392},{"year":2018,"month":1,"day":3,"energy_wh":4163},{"year":2018,"month":1,"day":4,"energy_wh":1207},{"year":2018,"month":1,"day":5,"energy_wh":3786},{"year":2018,"month":1,"day":6,"energy_wh":2170},{"year":2018,"month":1,"day":7,"energy_wh":4606},{"year":2018,"month":1,"day":8,"energy_wh":1560},{"year":2018,"month":1,"day":9,"energy_wh":2664},{"year":2018,"month":1,"day":10,"energy_wh":4567},{"year":2018,"month":1,"day":11,"energy_wh":3770},{"year":2018,"month":1,"day":12,"energy_wh":4956},{"year":2018,"month":1,"day":13,"energy_wh":481},{"year":2018,"month":1,"day":14,"energy_wh":4608},{"year":2018,"month":1,"day":15,"energy_wh":1538},{"year":2018,"month":1,"day":16,"energy_wh":304},{"year":2018,"month":1,"day":17,"energy_wh":976},{"year":2018,"month":1,"day":18,"energy_wh":3595},{"year":2018,"month":1,"day":19,"energy_wh":2125},{"year":2018,"month":1,"day":20,"energy_wh":1372},{"year":2018,"month":1,"day":21,"energy_wh":1403},{"year":2018,"month":1,"day":22,"energy_wh":3405},{"year":2018,"month":1,"day":23,"energy_wh":2924},{"year":2018,"month":1,"day":24,"energy_wh":2599},{"year":2018,"month":1,"day":25,"energy_wh":2839},{"year":2018,"month":1,"day":26,"energy_wh":1329},{"year":2018,"month":1,"day":27,"energy_wh":4928},{"year":2018,"month":1,"day":28,"energy_wh":4282},{"year":2018,"month":1,"day":29,"energy_wh":3392},{"year":2018,"month":1,"day":30,"energy_wh":4163},{"year":2018,"month":1,"day":31,"energy_wh":1207}],"err_code":0}}}'

var FebWatts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":2,"day":1,"energy":3.786},{"year":2018,"month":2,"day":2,"energy":2.17},{"year":2018,"month":2,"day":3,"energy":4.606},{"year":2018,"month":2,"day":4,"energy":1.56},{"year":2018,"month":2,"day":5,"energy":2.664},{"year":2018,"month":2,"day":6,"energy":4.567},{"year":2018,"month":2,"day":7,"energy":3.77},{"year":2018,"month":2,"day":8,"energy":4.956},{"year":2018,"month":2,"day":9,"energy":0.481},{"year":2018,"month":2,"day":10,"energy":4.608},{"year":2018,"month":2,"day":11,"energy":1.538},{"year":2018,"month":2,"day":12,"energy":0.304},{"year":2018,"month":2,"day":13,"energy":0.976},{"year":2018,"month":2,"day":14,"energy":3.595},{"year":2018,"month":2,"day":15,"energy":2.125},{"year":2018,"month":2,"day":16,"energy":1.372},{"year":2018,"month":2,"day":17,"energy":1.403},{"year":2018,"month":2,"day":18,"energy":3.405},{"year":2018,"month":2,"day":19,"energy":4.282},{"year":2018,"month":2,"day":20,"energy":3.392},{"year":2018,"month":2,"day":21,"energy":4.163},{"year":2018,"month":2,"day":22,"energy":1.207},{"year":2018,"month":2,"day":23,"energy":3.786},{"year":2018,"month":2,"day":24,"energy":2.17},{"year":2018,"month":2,"day":25,"energy":4.606},{"year":2018,"month":2,"day":26,"energy":1.56},{"year":2018,"month":2,"day":27,"energy":2.664},{"year":2018,"month":2,"day":28,"energy":4.567}],"err_code":0}}}'

var FebMWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":2,"day":1,"energy_wh":3786},{"year":2018,"month":2,"day":2,"energy_wh":2170},{"year":2018,"month":2,"day":3,"energy_wh":4606},{"year":2018,"month":2,"day":4,"energy_wh":1560},{"year":2018,"month":2,"day":5,"energy_wh":2664},{"year":2018,"month":2,"day":6,"energy_wh":4567},{"year":2018,"month":2,"day":7,"energy_wh":3770},{"year":2018,"month":2,"day":8,"energy_wh":4956},{"year":2018,"month":2,"day":9,"energy_wh":481},{"year":2018,"month":2,"day":10,"energy_wh":4608},{"year":2018,"month":2,"day":11,"energy_wh":1538},{"year":2018,"month":2,"day":12,"energy_wh":304},{"year":2018,"month":2,"day":13,"energy_wh":976},{"year":2018,"month":2,"day":14,"energy_wh":3595},{"year":2018,"month":2,"day":15,"energy_wh":2125},{"year":2018,"month":2,"day":16,"energy_wh":1372},{"year":2018,"month":2,"day":17,"energy_wh":1403},{"year":2018,"month":2,"day":18,"energy_wh":3405},{"year":2018,"month":2,"day":19,"energy_wh":4282},{"year":2018,"month":2,"day":20,"energy_wh":3392},{"year":2018,"month":2,"day":21,"energy_wh":4163},{"year":2018,"month":2,"day":22,"energy_wh":1207},{"year":2018,"month":2,"day":23,"energy_wh":3786},{"year":2018,"month":2,"day":24,"energy_wh":2170},{"year":2018,"month":2,"day":25,"energy_wh":4606},{"year":2018,"month":2,"day":26,"energy_wh":1560},{"year":2018,"month":2,"day":27,"energy_wh":2664},{"year":2018,"month":2,"day":28,"energy_wh":4567}],"err_code":0}}}'

var MarWatts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy":3.77},{"year":2018,"month":3,"day":2,"energy":4.956},{"year":2018,"month":3,"day":3,"energy":0.481},{"year":2018,"month":3,"day":4,"energy":4.608},{"year":2018,"month":3,"day":5,"energy":1.538},{"year":2018,"month":3,"day":6,"energy":0.304},{"year":2018,"month":3,"day":7,"energy":0.976},{"year":2018,"month":3,"day":8,"energy":3.595},{"year":2018,"month":3,"day":9,"energy":2.125},{"year":2018,"month":3,"day":10,"energy":1.372},{"year":2018,"month":3,"day":11,"energy":1.403},{"year":2018,"month":3,"day":12,"energy":3.405},{"year":2018,"month":3,"day":13,"energy":3.94},{"year":2018,"month":3,"day":14,"energy":3.77},{"year":2018,"month":3,"day":15,"energy":4.956},{"year":2018,"month":3,"day":16,"energy":0.481},{"year":2018,"month":3,"day":17,"energy":4.608},{"year":2018,"month":3,"day":18,"energy":1.538},{"year":2018,"month":3,"day":19,"energy":0.304},{"year":2018,"month":3,"day":20,"energy":1.56},{"year":2018,"month":3,"day":21,"energy":2.664},{"year":2018,"month":3,"day":22,"energy":4.567}],"err_code":0}}}'

var MarMWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy_wh":3770},{"year":2018,"month":3,"day":2,"energy_wh":4956},{"year":2018,"month":3,"day":3,"energy_wh":481},{"year":2018,"month":3,"day":4,"energy_wh":4608},{"year":2018,"month":3,"day":5,"energy_wh":1538},{"year":2018,"month":3,"day":6,"energy_wh":304},{"year":2018,"month":3,"day":7,"energy_wh":976},{"year":2018,"month":3,"day":8,"energy_wh":3595},{"year":2018,"month":3,"day":9,"energy_wh":2125},{"year":2018,"month":3,"day":10,"energy_wh":1372},{"year":2018,"month":3,"day":11,"energy_wh":1403},{"year":2018,"month":3,"day":12,"energy_wh":3405},{"year":2018,"month":3,"day":13,"energy_wh":3940},{"year":2018,"month":3,"day":14,"energy_wh":3770},{"year":2018,"month":3,"day":15,"energy_wh":4956},{"year":2018,"month":3,"day":16,"energy_wh":481},{"year":2018,"month":3,"day":17,"energy_wh":4608},{"year":2018,"month":3,"day":18,"energy_wh":1538},{"year":2018,"month":3,"day":19,"energy_wh":304},{"year":2018,"month":3,"day":20,"energy_wh":1560},{"year":2018,"month":3,"day":21,"energy_wh":2664},{"year":2018,"month":3,"day":22,"energy_wh":4567}],"err_code":0}}}'

var Day1Watts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy":3.77}],"err_code":0}}}'

var Day1MWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy_wh":3770}],"err_code":0}}}'

var Day2Watts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy":3.77},{"year":2018,"month":3,"day":2,"energy":4.956}],"err_code":0}}}'

var Day2MWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy_wh":3770},{"year":2018,"month":3,"day":2,"energy_wh":4956}],"err_code":0}}}'

var Day3Watts = '{"emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy":3.77},{"year":2018,"month":3,"day":2,"energy":4.956},{"year":2018,"month":3,"day":3,"energy":0.481}],"err_code":0}}}'

var Day3MWatts = '{"smartlife.iot.common.emeter":{"get_daystat":{"day_list":[{"year":2018,"month":3,"day":1,"energy_wh":3770},{"year":2018,"month":3,"day":2,"energy_wh":4956},{"year":2018,"month":3,"day":3,"energy_wh":481}],"err_code":0}}}'

var Null = '{"emeter":{"get_daystat":{"day_list":[{}],"err_code":0}}}'

var onWatts = '{"emeter":{"get_realtime":{"power":1234.567,"err_code":0}}}'

var offWatts = '{"emeter":{"get_realtime":{"power":0,"err_code":0}}}'